File: src/main/kotlin/br/com/ope/security/jwt/provider/JwtAuthenticationProvider.java
package br.com.ope.security.jwt.provider;

import br.com.ope.model.Usuario;
import br.com.ope.security.UserDetailsServiceImpl;
import br.com.ope.security.jwt.auth.JwtAuthenticationToken;
import br.com.ope.security.jwt.config.JwtSettings;
import br.com.ope.security.jwt.exceptions.UsuarioDesabilitadoException;
import br.com.ope.security.jwt.model.token.RawAccessJwtToken;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import org.apache.commons.lang3.StringUtils;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Component
@SuppressWarnings("unchecked")
public class JwtAuthenticationProvider implements AuthenticationProvider {
	private final JwtSettings jwtSettings;
    private final UserDetailsServiceImpl userDetailsService;
	
	public JwtAuthenticationProvider(JwtSettings jwtSettings, UserDetailsServiceImpl userDetailsService) {
		this.jwtSettings = jwtSettings;
		this.userDetailsService = userDetailsService;
	}
	
	@Override
	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
		RawAccessJwtToken rawAccessToken = (RawAccessJwtToken) authentication.getCredentials();

		if (StringUtils.isBlank(rawAccessToken.getToken())) {
			return new JwtAuthenticationToken(null, new ArrayList<>());
		}

		Jws<Claims> jwsClaims = rawAccessToken.parseClaims(jwtSettings.getTokenSigningKey());
		String subject = jwsClaims.getBody().getSubject();
		List<String> scopes = jwsClaims.getBody().get("scopes", List.class);
		Set<GrantedAuthority> authorities = scopes.stream()
				.map(SimpleGrantedAuthority::new)
				.collect(Collectors.toSet());
		
		Usuario context = new Usuario(subject, authorities);

        UserDetails usuario = userDetailsService.loadUserByUsername(subject);

        if (!usuario.isEnabled()) {
            throw new UsuarioDesabilitadoException("Usuario nao esta habilitado");
        }
		
		return new JwtAuthenticationToken(context, context.getAuthorities());
	}

	@Override
	public boolean supports(Class<?> authentication) {
		return (JwtAuthenticationToken.class.isAssignableFrom(authentication));
	}
}

File: src/main/kotlin/br/com/ope/security/jwt/provider/JwtAuthenticationProvider.kt
package br.com.ope.security.jwt.provider

import br.com.ope.model.Usuario
import br.com.ope.security.UserDetailsServiceImpl
import br.com.ope.security.jwt.auth.JwtAuthenticationToken
import br.com.ope.security.jwt.config.JwtSettings
import br.com.ope.security.jwt.exceptions.UsuarioDesabilitadoException
import br.com.ope.security.jwt.model.token.RawAccessJwtToken
import org.apache.commons.lang3.StringUtils
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Component
import java.util.*

@Component
class JwtAuthenticationProvider(private val jwtSettings: JwtSettings, private val userDetailsService: UserDetailsServiceImpl) : AuthenticationProvider {

    @Throws(AuthenticationException::class)
    override fun authenticate(authentication: Authentication): Authentication {
        val rawAccessToken = authentication.credentials as RawAccessJwtToken

        if (StringUtils.isBlank(rawAccessToken.token)) {
            return JwtAuthenticationToken(null, ArrayList())
        }

        val jwsClaims = rawAccessToken.parseClaims(jwtSettings.tokenSigningKey)
        val subject = jwsClaims.body.subject
        val scopes = jwsClaims.body.get("scopes", MutableList::class.java)
        val authorities : MutableSet<GrantedAuthority> = hashSetOf()
        for (scope in scopes) {
            authorities.add(SimpleGrantedAuthority(scope as String))
        }

        val context = Usuario(subject, authorities)

        val usuario = userDetailsService.loadUserByUsername(subject)

        if (!usuario.isEnabled) {
            throw UsuarioDesabilitadoException("Usuario nao esta habilitado")
        }

        return JwtAuthenticationToken(context, context.getAuthorities())
    }

    override fun supports(authentication: Class<*>): Boolean {
        return JwtAuthenticationToken::class.java.isAssignableFrom(authentication)
    }
}
